README.txt
Descartes
=========

Use Shapely_ or GeoJSON-like geometric objects as matplotlib paths and patches

.. image:: http://farm4.static.flickr.com/3662/4555372019_9bbed1f956_o_d.png
   :width: 800
   :height: 320

Requires: matplotlib, numpy, and optionally Shapely 1.2+.

Example::

  from matplotlib import pyplot
  from shapely.geometry import LineString
  from descartes import PolygonPatch
  
  BLUE = '#6699cc'
  GRAY = '#999999'
  
  def plot_line(ax, ob):
      x, y = ob.xy
      ax.plot(x, y, color=GRAY, linewidth=3, solid_capstyle='round', zorder=1)
  
  line = LineString([(0, 0), (1, 1), (0, 2), (2, 2), (3, 1), (1, 0)])
  
  fig = pyplot.figure(1, figsize=(10, 4), dpi=180)
  
  # 1
  ax = fig.add_subplot(121)
  
  plot_line(ax, line)
  
  dilated = line.buffer(0.5)
  patch1 = PolygonPatch(dilated, fc=BLUE, ec=BLUE, alpha=0.5, zorder=2)
  ax.add_patch(patch1)
  
  #2
  ax = fig.add_subplot(122)
  
  patch2a = PolygonPatch(dilated, fc=GRAY, ec=GRAY, alpha=0.5, zorder=1)
  ax.add_patch(patch2a)
  
  eroded = dilated.buffer(-0.3)
  
  # GeoJSON-like data works as well
  
  polygon = eroded.__geo_interface__
  # >>> geo['type']
  # 'Polygon'
  # >>> geo['coordinates'][0][:2]
  # ((0.50502525316941682, 0.78786796564403572), (0.5247963548222736, 0.8096820147509064))
  patch2b = PolygonPatch(polygon, fc=BLUE, ec=BLUE, alpha=0.5, zorder=2)
  ax.add_patch(patch2b)
  
  pyplot.show()


See also: examples/patches.py.

Descartes is not associated with the identically named and apparently defunct
project at http://descartes.sourceforge.net/.

.. _Shapely: http://gispython.org/lab/wiki/Shapely



CHANGES.txt
Changes
=======

Unless otherwise noted, issue and pull requests references are relative to
https://bitbucket.org/sgillies/descartes.

1.0.2 (2016-01-19)
------------------
- Slice coordinates to eliminate Z values of 3D geometries (PR #1).

1.0.1 (2013-11-15)
------------------
- Get long description in 'r' mode so setup works on Python 3.

1.0 (2010-04-26)
----------------
- Complete documentation and example.

0.1.2 (2010-04-09)
------------------
- Add support for the Python geo interface.

0.1.1 (2010-04-08)
------------------
- Add support for GeoJSON-like dicts.

0.1   (2010-04-07)
------------------
- Make path patches from Shapely polygons.


descartes/tests.py
from shapely.geometry import *
import unittest

from descartes.patch import PolygonPatch

class PolygonTestCase(unittest.TestCase):
    polygon = Point(0, 0).buffer(10.0).difference(
                MultiPoint([(-5, 0), (5, 0)]).buffer(3.0))
    def test_patch(self):
        patch = PolygonPatch(self.polygon)
        self.failUnlessEqual(str(type(patch)), 
            "<class 'matplotlib.patches.PathPatch'>")

class JSONPolygonTestCase(unittest.TestCase):
    polygon = Point(0, 0).buffer(10.0).difference(
                MultiPoint([(-5, 0), (5, 0)]).buffer(3.0))
    def test_patch(self):
        geo = self.polygon.__geo_interface__
        patch = PolygonPatch(geo)
        self.failUnlessEqual(str(type(patch)), 
            "<class 'matplotlib.patches.PathPatch'>")


descartes/patch.py
from matplotlib.patches import PathPatch
from matplotlib.path import Path
from numpy import asarray, concatenate, ones


class Polygon(object):
    # Adapt Shapely or GeoJSON/geo_interface polygons to a common interface
    def __init__(self, context):
        if hasattr(context, 'interiors'):
            self.context = context
        else:
            self.context = getattr(context, '__geo_interface__', context)
    @property
    def geom_type(self):
        return (getattr(self.context, 'geom_type', None)
                or self.context['type'])
    @property
    def exterior(self):
        return (getattr(self.context, 'exterior', None) 
                or self.context['coordinates'][0])
    @property
    def interiors(self):
        value = getattr(self.context, 'interiors', None)
        if value is None:
            value = self.context['coordinates'][1:]
        return value


def PolygonPath(polygon):
    """Constructs a compound matplotlib path from a Shapely or GeoJSON-like
    object"""
    this = Polygon(polygon)
    assert this.geom_type == 'Polygon'
    def coding(ob):
        # The codes will be all "LINETO" commands, except for "MOVETO"s at the
        # beginning of each subpath
        n = len(getattr(ob, 'coords', None) or ob)
        vals = ones(n, dtype=Path.code_type) * Path.LINETO
        vals[0] = Path.MOVETO
        return vals
    vertices = concatenate(
                    [asarray(this.exterior)] 
                    + [asarray(r) for r in this.interiors])
    codes = concatenate(
                [coding(this.exterior)] 
                + [coding(r) for r in this.interiors])
    return Path(vertices, codes)


def PolygonPatch(polygon, **kwargs):
    """Constructs a matplotlib patch from a Shapely or GeoJSON-like
    object with full support for polygon holes"""
    return PathPatch(PolygonPath(polygon), **kwargs)


examples/buffering.py
from matplotlib import pyplot
from shapely.geometry import LineString
from shapely.geometry.polygon import LinearRing

from descartes.patch import PolygonPatch

# 1
fig = pyplot.figure(1, figsize=(5, 4), dpi=180)
fig.set_frameon(True)
ax = fig.add_subplot(111)

a = LineString([(0, 0), (1, 1), (0, 2), (2, 2), (3, 1), (1, 0)])

dilated = a.buffer(0.5)
patch1 = PolygonPatch(dilated, facecolor='#99ccff', edgecolor='#6699cc')
ax.add_patch(patch1)

x, y = a.xy
ax.plot(x, y, color='#999999')

#ax.set_axis_off()
ax.set_xlim(-1, 4)
ax.set_ylim(-1, 3)
fig.savefig('buffering-1.png')

#2
fig2 = pyplot.figure(2, figsize=(5, 4), dpi=180)
fig2.set_frameon(True)
ax = fig2.add_subplot(111)

patch2a = PolygonPatch(dilated, facecolor='#cccccc', edgecolor='#999999')
ax.add_patch(patch2a)

eroded = dilated.buffer(-0.3)

patch2b = PolygonPatch(eroded, facecolor='#99ccff', edgecolor='#6699cc')
ax.add_patch(patch2b)

ax.set_xlim(-1, 4)
ax.set_ylim(-1, 3)
#ax.set_axis_off()
fig2.savefig('buffering-2.png')




examples/patches.py
from matplotlib import pyplot
from shapely.geometry import *

from descartes import PolygonPatch


# Create a matplotlib figure
fig = pyplot.figure(num=1, figsize=(10, 4), dpi=180)

# Create a polygon with 2 holes using buffer and difference methods
polygon = Point(0, 0).buffer(10.0).difference(
    MultiPoint([(-5, 0), (5, 0)]).buffer(3.0))

# 1
# Create a subplot
ax = fig.add_subplot(121)

# Make the polygon into a patch and add it to the subplot
patch = PolygonPatch(polygon, facecolor='#cccccc', edgecolor='#999999')
ax.add_patch(patch)

# Fit the figure around the polygon's bounds, render, and save
minx, miny, maxx, maxy = polygon.bounds
w, h = maxx - minx, maxy - miny
ax.set_xlim(minx - 0.2*w, maxx + 0.2*w)
ax.set_ylim(miny - 0.2*h, maxy + 0.2*h)
ax.set_aspect(1)

# 2
# Create a subplot
ax = fig.add_subplot(122)

# Turn the GeoJSON-ish dict form of the polygon from #1 into a patch
geo = polygon.__geo_interface__
patch = PolygonPatch(geo, facecolor='#cccccc', edgecolor='#999999')
ax.add_patch(patch)

# Fit the figure around the polygon's bounds, render, and save
minx, miny, maxx, maxy = polygon.bounds
w, h = maxx - minx, maxy - miny
ax.set_xlim(minx - 0.2*w, maxx + 0.2*w)
ax.set_ylim(miny - 0.2*h, maxy + 0.2*h)
ax.set_aspect(1)

fig.savefig('patches.png')

